### DISCUSSION TITLE
How could I know all ws connections were  actually closed?

### DISCUSSION THREAD

**Discussion by pengweichu** [uNetworking/uWebSockets#1120] [Q&A] [answered]
Below is my code, since the loop->defer is async, after I called removeAllClients, how can I determine all clients are actually closed? Thanks

```
    void removeAllClients()
    {
	std::unique_lock<std::mutex> lock(mMutex);
        for (auto & it : mClients)
        {
            auto ws = it.second;
			mLoop->defer([ws]()
				{
					ws->close();
				});
        }

        mClients.clear();
    }
```

**Comment by ghost** [comment, marked as answer]
You can read the docs; https://unetworking.github.io/uWebSockets.js/generated/interfaces/websocket.html#close

That code is very inefficient as you're making one defer per socket rather than one defer in total.
